### ISSUE TITLE
Improvement - Separate chameleon attributes into new data classes

### ISSUE THREAD

**Issue by LeeOhHyung** [sangcomz/Chameleon#21]
```kotlin
[ChameleonAttr.kt]
data class ChameleonAttr(var emptyText: String,
                         var emptyTextColor: Int,
                         var emptyTextSize: Float,
                         var emptyTextGravity: Int,
                         var emptySubText: String,
                         var emptySubTextColor: Int,
                         var emptySubTextSize: Float,
                         var emptySubTextGravity: Int,
                         var emptyDrawable: Drawable?,
                         var emptyButtonText: String,
                         var emptyButtonTextColor: Int,
                         var emptyButtonTextSize: Float,
                         var emptyButtonBackgroundColor: Int,
                         var useEmptyButton: Boolean,
                         var errorText: String,
                         var errorTextColor: Int,
                         var errorTextSize: Float,
                         var errorTextGravity: Int,
                         var errorSubText: String,
                         var errorSubTextColor: Int,
                         var errorSubTextSize: Float,
                         var errorSubTextGravity: Int,
                         var errorDrawable: Drawable?,
                         var errorButtonText: String,
                         var errorButtonTextColor: Int,
                         var errorButtonTextSize: Float,
                         var errorButtonBackgroundColor: Int,
                         var useErrorButton: Boolean,
                         var progressDrawable: Drawable?,
                         var useProgressBackground: Boolean,
                         var progressBackgroundColor: Int,
                         var isShowProgressWhenContentState: Boolean,
                         var isLargeProgress: Boolean,
                         var defaultState: Chameleon.STATE,
                         var stateImageBottomMargin:Float,
                         var subTitleTextTopMargin:Float,
                         var buttonTopMargin:Float,
                         val titleFontFamily:Int,
                         val subTitleFontFamily:Int,
                         val buttonFontFamily:Int
)

[chameleon/Chameleon.kt]
ChameleonAttr(
                it.getString(R.styleable.Chameleon_emptyText) ?: "empty",
                it.getColor(
                    R.styleable.Chameleon_emptyTextColor,
                    ContextCompat.getColor(context, R.color.colorTitleText)
                ),
                it.getDimension(
                    R.styleable.Chameleon_emptyTextSize,
                    context.resources.getDimension(R.dimen.title_text_size)
                ),
                it.getInt(R.styleable.Chameleon_emptyTextGravity, 0),
                it.getString(R.styleable.Chameleon_emptySubText) ?: "empty content",
                it.getColor(
                    R.styleable.Chameleon_emptySubTextColor,
                    ContextCompat.getColor(context, R.color.colorSubText)
                ),
                it.getDimension(
                    R.styleable.Chameleon_emptySubTextSize,
                    context.resources.getDimension(R.dimen.sub_text_size)
                ),
                it.getInt(R.styleable.Chameleon_emptySubTextGravity, 0),
                it.getResourceId(
                    R.styleable.Chameleon_emptyDrawable,
                    R.drawable.ic_chameleon_empty
                ).getDrawable(context),
                it.getString(R.styleable.Chameleon_emptyButtonText) ?: "retry",
                it.getColor(
                    R.styleable.Chameleon_emptyButtonTextColor,
                    ContextCompat.getColor(context, R.color.colorTitleText)
                ),
                it.getDimension(
                    R.styleable.Chameleon_emptyButtonTextSize,
                    context.resources.getDimension(R.dimen.title_text_size)
                ),
                it.getColor(
                    R.styleable.Chameleon_emptyButtonBackgroundColor,
                    ContextCompat.getColor(context, R.color.colorSubText)
                ),
                it.getBoolean(R.styleable.Chameleon_useEmptyButton, false),
                it.getString(R.styleable.Chameleon_errorText) ?: "error",
                it.getColor(
                    R.styleable.Chameleon_errorTextColor,
                    ContextCompat.getColor(context, R.color.colorTitleText)
                ),
                it.getDimension(
                    R.styleable.Chameleon_errorTextSize,
                    context.resources.getDimension(R.dimen.title_text_size)
                ),
                it.getInt(R.styleable.Chameleon_errorTextGravity, 0),
                it.getString(R.styleable.Chameleon_errorSubText) ?: "error content",
                it.getColor(
                    R.styleable.Chameleon_errorSubTextColor,
                    ContextCompat.getColor(context, R.color.colorSubText)
                ),
                it.getDimension(
                    R.styleable.Chameleon_errorSubTextSize,
                    context.resources.getDimension(R.dimen.sub_text_size)
                ),
                it.getInt(R.styleable.Chameleon_errorSubTextGravity, 0),
                it.getResourceId(
                    R.styleable.Chameleon_errorDrawable,
                    R.drawable.ic_chameleon_error
                ).getDrawable(context),
                it.getString(R.styleable.Chameleon_errorButtonText) ?: "retry",
                it.getColor(
                    R.styleable.Chameleon_errorButtonTextColor,
                    ContextCompat.getColor(context, R.color.colorTitleText)
                ),
                it.getDimension(
                    R.styleable.Chameleon_errorButtonTextSize,
                    context.resources.getDimension(R.dimen.title_text_size)
                ),
                it.getColor(
                    R.styleable.Chameleon_errorButtonBackgroundColor,
                    ContextCompat.getColor(context, R.color.colorTitleText)
                ),
                it.getBoolean(R.styleable.Chameleon_useErrorButton, false),
                it.getDrawable(R.styleable.Chameleon_progressDrawable),
                it.getBoolean(R.styleable.Chameleon_useProgressBackground, false),
                it.getColor(
                    R.styleable.Chameleon_progressBackgroundColor,
                    ContextCompat.getColor(context, R.color.colorLoadingBackground)
                ),
                it.getBoolean(R.styleable.Chameleon_isShowContentWhenLoadingState, false),
                it.getBoolean(R.styleable.Chameleon_isLargeProgress, false),
                stateFromInt(it.getInt(R.styleable.Chameleon_defaultChameleonState, -1)),
                it.getDimension(
                    R.styleable.Chameleon_stateImageBottomMargin,
                    4.DP(context)
                ),
                it.getDimension(
                    R.styleable.Chameleon_SubTitleTextTopMargin,
                    4.DP(context)
                ),
                it.getDimension(
                    R.styleable.Chameleon_ButtonTopMargin,
                    16.DP(context)
                ),
                it.getResourceId(R.styleable.Chameleon_titleFontFamily, -1),
                it.getResourceId(R.styleable.Chameleon_subTitleFontFamily, -1),
                it.getResourceId(R.styleable.Chameleon_buttonFontFamily, -1)
            )
```
#### Discussion
 - If you want to create new ChameleonAttr object, too many parameters make
the code difficult to read and increase line of code.
 - I think there are too many parameters in Chameleon data class. 
 - So. How about separate attributes into smaller data class ??

#### for example. Combine similar attributes to new data class.
 - EmtpyAttr
 - ErrorAttr
 - progressSettingBundle
 - ChameleonMargins
 - ChameleonFontFamily

 #### Here is example code.
```kotlin
data class ChameleonAttr(val emptyAttr: EmptyAttr,
                         val errorAttr: ErrorAttr,
                         val progressSettingBundle: ProgressSettingBundle,
                         var defaultState: Chameleon.STATE,
                         val chameleonMargins: ChameleonMargins,
                         val chameleonFontFamily: ChameleonFontFamily)
```

**Comment by sangcomz** [comment]
It seems like a good way to make changes for readability!
Segmentation criteria are important.

Thank you for your good opinion. 😃

**Comment by LeeOhHyung** [comment]
I have changed some codes, but it's inefficient than I thought.
Because, when I the `ChameleonAttr` class, I had to  change many parts of the `Chameleon` class.

I'm not sure if chameleonAttr is necessarily effective in tying it together into common parts.
and if loot at the `activity_main.xml` of the sample app, adding the layout attributes one by one.
I think it is great advantage to be able to add attributes in this detail. If grouped into common part, cannot use this.

So. Group into common part is inefficient
